src/utils.cpp

```diff
@@ -195,6 +199,128 @@ bool is_lit(const Term & l, const Sort & boolsort)
   return false;
 }
 
+void cnf_to_dimacs(Term cnf, std::ostringstream & y)
+{
+  if (cnf->is_value() && cnf->to_string() == "true")
+  {  // empty cnf formula
+    y << "p cnf 0 0\n";
+    return;
+  }
+  TermVec vecs({ cnf });
+  TermVec vecs2;
+  // This while loop separate the clauses, vecs2 will contain all clauses
+  // because every smt::and op will be eliminated. This happens because until
+  // smt::and op is not detected that term is added back to vecs, and as no term
+  // with smt::or as the primOp is touched all clauses shall be separated and be
+  // intact
+  while (!vecs.empty())
+  {
+    Term t = vecs.back();
+    vecs.pop_back();
+    smt::Op op = t->get_op();
+    if (op.prim_op == smt::And)
+    {
+      for (auto u : t)
+      {
+        vecs.push_back(u);
+      }
+    }
+    else
+    {
+      vecs2.push_back(t);
+    }
+  }
+  // Storing literals from each clause, each vector in literals will contain the
+  // literals from a clause
+  std::vector<std::vector<Term>> literals;
+
+  for (auto u : vecs2)
+  {
+    std::vector<Term>add;
+    std::vector<Term>le({u});
+    while (!le.empty())
+    {  // This while loop functions in the same way as above and eliminates
+       // smt::or by separating the literals
+      Term t=le.back();
+      le.pop_back();
+      smt::Op op=t->get_op();
+      if(op.prim_op==smt::Or){
+        for(auto u:t){
+          le.push_back(u);
+        }
+      }
+      else{
+        add.push_back(t);
+      }
+    }
+    literals.push_back(add);
+   }
+
+   std::map<std::string, int>
+       ma;  // This map will create a mapping from symbols to distinct
+            // contiguous integer values.
+   int ptr = 0;  // pointer to store the next integer used in mapping
+
+   for(auto u:literals){
+     for (auto uu : u)
+     {  // Using literals from all the clauses to create the mapping
+       if (uu->is_value())
+       {  // For an empty clause, which will just contain the term "false"
+       }
+       else if (uu->is_symbolic_const())
+       {  // A positive literal
+         if (ma.find(uu->to_string()) == ma.end())
+         {  // Checking if symbol is absent in the mapping done till now
+           ptr++;
+           ma[uu->to_string()] = ptr;
+         }
+       }
+       else
+       {  // A negative literal
+         Term t = (*(uu->begin()));
```

Comment:
Can you use `uu[0]`? I think this what we usually do in the code, but correct me if I am wrong.

Reply:
I tried uu[0] right now and got a compilation error, I picked this notation from is_lit() function in the same file :)

Reply:
:) Well I guess we don't usually use it in the code then!